internal/crypto/internal_ca.go

```diff
@@ -0,0 +1,209 @@
+	return ca.Config
+}
+
+func (ca *internalCA) EnsureServerCertificate(certFile, keyFile string, hostnames []string, expireDays int) (*TLSCertificateConfig, bool, error) {
+	certConfig, err := GetServerCert(certFile, keyFile, hostnames)
+	if err != nil {
+		certConfig, err = ca.MakeAndWriteServerCert(certFile, keyFile, hostnames, expireDays)
+		return certConfig, true, err
+	}
+
+	return certConfig, false, nil
+}
+
+func (ca *internalCA) MakeAndWriteServerCert(certFile, keyFile string, hostnames []string, expireDays int) (*TLSCertificateConfig, error) {
+	server, err := ca.MakeServerCert(hostnames, expireDays)
+	if err != nil {
+		return nil, err
+	}
+	if err := server.WriteCertConfigFile(certFile, keyFile); err != nil {
+		return server, err
+	}
+	return server, nil
+}
+
+func (ca *internalCA) MakeServerCert(hostnames []string, expiryDays int, fns ...CertificateExtensionFunc) (*TLSCertificateConfig, error) {
+	if len(hostnames) < 1 {
+		return nil, fmt.Errorf("at least one hostname must be provided")
+	}
+
+	serverPublicKey, serverPrivateKey, publicKeyHash, _ := NewKeyPairWithHash()
+
+	now := time.Now()
+	serverTemplate := &x509.Certificate{
+		Subject: pkix.Name{CommonName: hostnames[0]},
+
+		SignatureAlgorithm: x509.ECDSAWithSHA256,
+
+		NotBefore:    now.Add(-1 * time.Second),
+		NotAfter:     now.Add(time.Duration(expiryDays) * 24 * time.Hour),
+		SerialNumber: big.NewInt(1),
+
+		KeyUsage:              x509.KeyUsageKeyEncipherment | x509.KeyUsageDigitalSignature,
+		ExtKeyUsage:           []x509.ExtKeyUsage{x509.ExtKeyUsageServerAuth},
+		BasicConstraintsValid: true,
+
+		AuthorityKeyId: ca.Config.Certs[0].SubjectKeyId,
+		SubjectKeyId:   publicKeyHash,
+	}
+	serverTemplate.IPAddresses, serverTemplate.DNSNames = oscrypto.IPAddressesDNSNames(hostnames)
+	for _, fn := range fns {
+		if err := fn(serverTemplate); err != nil {
+			return nil, err
+		}
+	}
+
+	serverCrt, err := ca.signCertificate(serverTemplate, serverPublicKey)
+	if err != nil {
+		return nil, err
+	}
+	server := &TLSCertificateConfig{
+		Certs: append([]*x509.Certificate{serverCrt}, ca.Config.Certs...),
+		Key:   serverPrivateKey,
+	}
+	return server, nil
+}
+
+func (ca *internalCA) signCertificate(template *x509.Certificate, requestKey crypto.PublicKey) (*x509.Certificate, error) {
+	// Increment and persist serial
+	serial, err := ca.SerialGenerator.Next(template)
+	if err != nil {
+		return nil, err
+	}
+	template.SerialNumber = big.NewInt(serial)
+	return signCertificate(template, requestKey, ca.Config.Certs[0], ca.Config.Key)
+}
+
+func (ca *internalCA) EnsureClientCertificate(certFile, keyFile string, subjectName string, expireDays int) (*TLSCertificateConfig, bool, error) {
+	certConfig, err := GetClientCertificate(certFile, keyFile, subjectName)
+	if err != nil {
+		certConfig, err = ca.MakeClientCertificate(certFile, keyFile, subjectName, expireDays)
+		return certConfig, true, err // true indicates we wrote the files.
```

Comment:
should we document the interface with what the values of this bool mean?

Reply:
This bool as undocumented. It was showing if the CA was generated or loaded from storage. Sorry, missed that one. I can add the relevant documentation.